### DISCUSSION TITLE
Created a custom GLSPDiagramManager and it fails to initialize

### DISCUSSION THREAD

**Discussion by dmm9** [eclipse-glsp/glsp#887] [Q&A] [answered]
Starting from the TaskList (java-emf-theia) example, I am trying to modify how the theia widget is created. In particular, I am trying to open the widget using a theia command from the command register. For that purpose, I have extended the class `GLSPDiagramManager` and created the bindings as indicated in the documentation of glsp-theia-container-module.ts. It compiles. However, the theia IDE does not load on the browser at all and I get the console error `Error: Ambiguous match found for serviceIdentifier: MyGLSPDiagramManager`. Searching on internet, it seems an issue with the bindings, but I don't know how to fix it.

Q1: Where is the mistake?
Q2: How can I open the widget from the command's execute method?

This is my frontend module:
```typescript
export class TaskListTheiaFrontendModule extends GLSPTheiaFrontendModule implements CommandContribution {
    readonly diagramLanguage = TaskListLanguage;

    bindDiagramConfiguration(context: ContainerContext): void {
        context.bind(DiagramConfiguration).to(TaskListDiagramConfiguration);
    }

    registerCommands(commands: CommandRegistry): void {
        commands.registerCommand(
            {
                id: 'tasklist-view',
                label: 'Task List'
            } , {
                execute: () => Function.prototype() // TBD
        });
    }

    override configureDiagramManager(context: ContainerContext): void {
        context.bind(MyGLSPDiagramManager).to(MyGLSPDiagramManager).inSingletonScope();
        registerDiagramManager(context.bind, MyGLSPDiagramManager);
    }
}
```
The bindings follow the documentation of `configureDiagramManager()`, which reads:
```typescript
/**
     * Configures the bindings for the diagram manager of the diagram integration. A {@link ConfigurableGLSPDiagramManager}
     * is bound to a generated service identifer in singleton scope and then additional bindings for this service
     * identifer are registered. Can be overwritten in subclasses to provide a custom binding.
     *
     *  For example:
     * ```typescript
     * context.bind(MyDiagramManager).to(MyDiagramManager).inSingletonScope();
     * registerDiagramManager(context.bind, MyDiagramManager);
     * ```
     * @param context the container context
     */
```
And `MyGLSPDiagramManager` is currently a carbon copy of `ConfigurableGLSPDiagramManager`.

**Comment by tortmayr** [comment, marked as answer]
Hi @dmm9 ,

>Starting from the TaskList (java-emf-theia) example, I am trying to modify how the theia widget is created. In particular, I am trying to open the widget using a theia command from the command register. For that purpose, I have extended the class GLSPDiagramManager and created the bindings as indicated in the documentation of glsp-theia-container-module.ts. It compiles. However, the theia IDE does not load on the browser at all and I get the console error Error: Ambiguous match found for serviceIdentifier: MyGLSPDiagramManager. Searching on internet, it seems an issue with the bindings, but I don't know how to fix it.
>
>Q1: Where is the mistake?

It looks like the documentation of `configureDiagramManager` is outdated.The  [registerDiagramManager](https://github.com/eclipse-glsp/glsp-theia-integration/blob/256b59459dcd4a166204f3f6f480dcb1d1c1e83a/packages/theia-integration/src/browser/diagram/glsp-diagram-manager.ts#L36) has a `bindToSelf` parameter which defaults to true.
That's why you end up with the `ambiguous match ...` error.
To solve this you can either remove your manual binding or set the parameter to false:
```ts
override configureDiagramManager(context: ContainerContext): void {
        registerDiagramManager(context.bind, MyGLSPDiagramManager);
    }
```
or

```ts
override configureDiagramManager(context: ContainerContext): void {
        context.bind(MyGLSPDiagramManager).to(MyGLSPDiagramManager).inSingletonScope();
        registerDiagramManager(context.bind, MyGLSPDiagramManager,false);
    }
 ```

>  Q2: How can I open the widget from the command's execute method?

You can use Theia's [OpenerService](https://github.com/eclipse-theia/theia/blob/ec7ed046b56f803490c85363ea47a09967cd57ff/packages/core/src/browser/opener-service.ts#L94) for this.
If you need an example, we already to something similar in our workflow dev-example:
https://github.com/eclipse-glsp/glsp-theia-integration/blob/1889d42d5a31e411d5bbcdd6d35c53b195c38a42/examples/workflow-theia/src/browser/diagram/workflow-diagram-readonly-view.ts#L46
(If you omit the `openerOptions` this should be pretty close to what you are trying to achieve)


 
 > And MyGLSPDiagramManager is currently a carbon copy of ConfigurableGLSPDiagramManager.

Please be carefull with the useage of  [`ConfigurableGLSPDiagramManager`](https://github.com/eclipse-glsp/glsp-theia-integration/blob/256b59459dcd4a166204f3f6f480dcb1d1c1e83a/packages/theia-integration/src/browser/glsp-theia-container-module.ts#L246). It is an internal class and the code might not be reusable 1 to 1 in your custom implementation. 
In this case, if you have a carbon copy you have to make sure that you call the `doConfigure` method because it will not be invoked automatically.

**Reply by dmm9** [reply]
You have made my day with both answers 😍 

> To solve this you can either remove your manual binding or set the parameter to false

That was it! I have set the parameter to false. It is not evident at first sight that there is a double bind of MyGLSPDiagramManager. And the documentation, as you noticed, doesn't help.

> If you need an example, we already to something similar in our workflow dev-example

This example is also very convenient for my use case and has helped me go further. 🚀

Thanks!!
